contracts/CDPLeverage.sol

```diff
@@ -0,0 +1,643 @@
+        uint256 currPricePeth = wmul(currPriceEth,wdiv(weth2peth(1*WAD),(1*WAD)));    // DAI/PETH In WAD
+
+        // ensure that the price floor is less than the current price of eth
+        require(_priceFloorETH < currPriceEth, "Price floor should not be less than the Dai/Eth price feed.");
+        require(0 < msg.value, "Ether is required to open a position.");
+        require(investors[msg.sender].principalETH == 0, "Previous position must be liquidated before opening a new one");
+
+        // calculate collateralization ratio from price floor
+        uint256 collatRatio = wdiv(wmul(currPriceEth, makerLR),_priceFloorETH);
+
+        //Add information about investor to database
+        Investor memory sender;
+        sender.index = investorAddresses.push(msg.sender) - 1;
+        sender.principalETH = msg.value;
+        sender.layers = _layers;
+        sender.collatRatio = collatRatio;
+        sender.priceFloor = _priceFloorETH;
+        sender.purchPrice = estimatePurchPrice(
+            sender.principalETH,
+            sender.collatRatio,
+            sender.layers
+        );
+
+        uint256 wethAmount;
+        uint256 pethAmount;
+        uint256 daiAmount;
+
+        IWETH(weth).deposit.value(msg.value)();
+        sender.cdpID = tub.open();
+
+        (daiAmount, pethAmount) = joinLockDraw(
+            sender.cdpID,
+            sender.principalETH,
+            currPricePeth,
+            sender.collatRatio
+        );
+
+        //trade DAI for WETH, convert WETH to PETH, and reinvest into CDP for # of layers
+        for (uint256 i = 0; i < sender.layers - 1; i++) {
+            wethAmount = marketBuy(weth, dai, daiAmount);
+            (daiAmount, pethAmount) = joinLockDraw(
+                sender.cdpID,
+                wethAmount,
+                currPricePeth,
+                sender.collatRatio
+            );
+
+        }
+
+        //final variable and struct assigment
+        sender.daiAmountFinal = daiAmount;
+        sender.drawnDai = wdiv(tub.tab(sender.cdpID),ray2wad(tub.chi()));
+        investors[msg.sender] = sender;
+
+        emit OpenPosition(msg.sender, msg.value, sender.purchPrice, _layers);
+    }
+
+
+    /**
+    * @dev Liquidate - Liquidates the sender's position and returns ETH
+    * Possible outcomes:
+    * 1. USD/ETH drops below sender's price floor
+    * 2. USD/ETH appreciates relative to sender's purchase price
+    * 3. USD/ETH deppreciates relative to sender's purchase price
+    * Sender must send ETH with function call to cover the governance and stability fee with use of the Maker platform
+    * Any excess ETH will be returned with the closing of the position
+    */
+    function liquidate()
+        payable
+        onlyInvestor
+        public
+    {
+        uint256 currPriceEth = uint256(pip.read());                          // DAI/WETH In WAD
+
+        Investor memory sender = investors[msg.sender];
+
+        uint256 releasedPeth;
```

Comment:
This is only used within the else block, can we move it in there?

Reply:
Yea, I see how it would save gas 👍